rfcs/react-components/convergence/event-handlers-event-type.md

```diff
@@ -0,0 +1,198 @@
+    | EventData<'scroll', Event>
+  ) & {
+    open: boolean;
+  };
+
+  type PopoverProps = {
+    /**
+     * @deprecated Use onOpenChange2 instead.
+     */
+    onOpenChange?: (e: OpenPopoverEvents, data: OnOpenChangeData) => void;
+
+    onOpenChange2?: EventHandler<OnOpenChangeData>;
+  };
+}
+
+// ======= For a new component =======
+{
+  type MyComponentElement = HTMLElement;
+
+  type OnSomeEventData = EventData<'click', React.MouseEvent<MyComponentElement>> & {
+    open: boolean;
+  };
+  // If one day we need to add more events, we can just add them to the union:
+  // type OnSomeEventData =
+  //   | EventData<'click', React.MouseEvent<MyComponentElement>>
+  //   | (EventData<'focus', React.FocusEvent<MyComponentElement>> & {
+  //       open: boolean;
+  //     });
+
+  type SomeProps = {
+    onSomeEvent?: EventHandler<OnSomeEventData>;
+  };
+}
+```
+
+- Pros 👍:
+
+  1. Ensures correct typing and it can accommodate future, unforeseen event types.
+  2. Currently we use unions to type events in cases where a callback may be triggered by multiple events. Consumers will need to use type assertion or type predict to use a strongly-typed event. But this solution provides an easier way:
+
+     ```ts
+     {
+       // Current way:
+       function isMouseEvent(e: MouseEvent | TouchEvent): e is MouseEvent {
+         return e instanceof MouseEvent; // This type check is just for demo purpose. There are issues with using `instanceof` in multi-window scenarios: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/instanceof#instanceof_and_multiple_realms
+       }
+       const onOpenChange = (e: MouseEvent | TouchEvent) => {
+         console.log(e.clientX); // ❌ type error - Property 'clientX' does not exist on type 'TouchEvent'
+         if (isMouseEvent(e)) {
+           console.log(e.clientX); // ✅
+         }
+       };
+     }
+
+     {
+       // Proposed way:
+       type OnSomeEventData =
+         | EventData<'click', React.MouseEvent<MyComponentElement>>
+         | (EventData<'focus', React.FocusEvent<MyComponentElement>> & {
+             open: boolean;
+           });
+       const onOpenChange = (_e: Event, data: OnSomeEventData) => {
+         console.log(data.event?.currentTarget); // base event data can be used without checking data.type
+         console.log(data.event?.clientX); // Typescript will give an error here as expected
+         if (data.type === 'click') {
+           console.log(data.event.clientX); // ✅ this won't blow as `data.type` verification will ensure `data.event` is a mouse event
+         }
+       };
+     }
+     ```
+
+- Cons 👎:
+  1. The complexity of the callback signature increases. Users need to discern the differences between the event argument and the events within the data argument.
+  2. Code bloat - as there are many callbacks, maintaining both versions of callbacks can become cumbersome. This is a problem that can only be solved in v10.
```

Comment:
(2) is less of an issue if we go with my suggestion of only deprecating/replacing the events on an as-needed basis.